### PR TITLE
Rewrite

### PR DESCRIPTION
**Pull request by StarArawn**
Closes: #170 
Closes: #205
Closes: #140
Closes: #121
Closes: #71

### CODE REVIEW

File: examples/random_map.rs [hunk 1]
```diff
@@ -1,61 +1,61 @@
...
+                .insert(LastUpdate::default())
+                .insert(TileVisible(true))
+                .id();
+            tile_storage.set(&tile_pos, Some(tile_entity));
         }
-        commands
-            .entity(tile_entity.unwrap())
-            .insert(LastUpdate::default());
-    });
+    }
```
Comment (modulozero) [inline examples/random_map.rs:39]: 🙏🏻

To the completely ignorant of game dev person that I am, this is so much more intuitive. Can't wait for the post-rewrite version.

File: src/tiles/mod.rs [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+}
+
+#[derive(Bundle, Default, Clone, Copy, Debug)]
+pub struct TileBundle {
+    pub position: TilePos2d,
+    pub texture: TileTexture,
+    pub tilemap_id: TilemapId,
+    pub visible: TileVisible,
+    pub flip: TileFlip,
+}
```
Comment (Trouv) [inline src/tiles/mod.rs:69]: Is there any play to support colors on tiles, like the old `Tile` component?

File: src/tiles/mod.rs [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+
+use bevy::{
+    math::UVec2,
+    prelude::{Bundle, Color, Component},
+};
+pub use storage::*;
+
+use crate::map::TilemapId;
+
+#[derive(Component, Default, Clone, Copy, Debug)]
```
Comment (Trouv) [inline src/tiles/mod.rs:11]:
```suggestion
#[derive(Component, Default, Clone, Copy, Debug, Hash)]
```
My plugin makes liberal use of `HashMap<TilePos>`

**Review by Trouv — CHANGES_REQUESTED** [review]

File: src/tiles/mod.rs (hunk 2 above)
Comment (Trouv) [reply src/tiles/mod.rs]: Thanks for implementing this!

**Comment by rparrett** [comment]
FWIW, I migrated a project to the rewrite branch and didn't have any issues. But I'm pretty much just loading a single static Tiled map.

**Comment by StarArawn** [comment]
> FWIW, I migrated a project to the rewrite branch and didn't have any issues. But I'm pretty much just loading a single static Tiled map.

That's great news! Thanks for letting me know! 🙂

File: src/tiles/mod.rs (hunk 3 above)
Comment (StarArawn) [reply src/tiles/mod.rs]: This is done.

**Comment by NeoRaider** [comment]
@StarArawn I think commit https://github.com/StarArawn/bevy_ecs_tilemap/pull/205/commits/926ceea43a617c379097819b3e0953c565be47d4 would still be a good improvement for the rewrite. Do you want to cherry-pick it, or should I open a new PR?

**Comment by StarArawn** [comment]
So, according to one of Bevy's rendering devs that import is going to be removed. It was suggested to just keep what I have for right now.

**Comment by NeoRaider** [comment]
I see! In that case the struct definition should be updated to https://github.com/bevyengine/bevy/blob/6752c9c59b3b81168ea6607767cbb519e5127976/crates/bevy_sprite/src/mesh2d/mesh2d_view_types.wgsl to match the Rust side of ViewUniform (this shader import was actually broken in the Bevy 0.8 release and fixed in a subsequent commit)

**Comment by StarArawn** [comment]
> I see! In that case the struct definition should be updated to https://github.com/bevyengine/bevy/blob/6752c9c59b3b81168ea6607767cbb519e5127976/crates/bevy_sprite/src/mesh2d/mesh2d_view_types.wgsl to match the Rust side of ViewUniform (this shader import was actually broken in the Bevy 0.8 release and fixed in a subsequent commit)

Ideally I suppose we should use bevy's import. Until we can I think copying it into our shader is probably best.

**Pull request merged**